dspy/predict/code_act.py

```diff
@@ -0,0 +1,121 @@
+            act(n=5) # 120
+            ```
+        """
+        self.signature = ensure_signature(signature)
+        self.max_iters = max_iters
+
+        tools = [t if isinstance(t, Tool) else Tool(t) for t in tools]
+        if any(
+            not inspect.isfunction(tool.func) for tool in tools
+        ):
+            raise ValueError("CodeAct only accepts functions and not callable objects.")
+        tools = {tool.name: tool for tool in tools}
+
+        instructions = self._build_instructions(self.signature, tools)
+
+        codeact_signature = (
+            dspy.Signature({**self.signature.input_fields}, "\n".join(instructions))
+            .append("trajectory", dspy.InputField(), type_=str)
+            .append("generated_code", dspy.OutputField(desc="python code that answers the question"), type_=str)
+            .append("finished", dspy.OutputField(desc="a boolean flag to determine if the process is done"), type_=bool)
+        )
+
+        extract_signature = dspy.Signature(
+            {**self.signature.input_fields, **self.signature.output_fields},
+            self.signature.instructions,
+        ).append("trajectory", dspy.InputField(), type_=str)
+
+        self.tools: dict[str, Tool] = tools
+        self.codeact = dspy.Predict(codeact_signature)
+        self.extractor = dspy.ChainOfThought(extract_signature)
+        # It will raises exception when dspy cannot find available deno instance by now.
+        self.interpreter = PythonInterpreter()
+    
+    def _build_instructions(self, signature, tools):
+        instructions = [f"{signature.instructions}\n"] if signature.instructions else []
+        inputs = ", ".join([f"`{k}`" for k in signature.input_fields.keys()])
+        outputs = ", ".join([f"`{k}`" for k in signature.output_fields.keys()])
+
+        instructions.append(
+            f"You are an intelligent agent. For each episode, you will receive the fields {inputs} as input.\n"
+            f"Your goal is to generate executable Python code that collects any necessary information for producing {outputs}.\n"
+            "For each iteration, you will generate a code snippet that either solves the task or progresses towards the solution.\n"
+            "Ensure any output you wish to extract from the code is printed to the console. The code should be enclosed in a fenced code block.\n"
+            f"When all information for producing the outputs ({outputs}) are available to be extracted, mark `finished=True` besides the final Python code.\n"
+            "You have access to the Python Standard Library and the following functions:"
+        )
+
+        for idx, tool in enumerate(tools.values()):
+            instructions.append(f"({idx + 1}) {tool}")
+        
+        return instructions
+
+    def forward(self, **kwargs):
+        # Define the tool funcitons in the interpreter
+        for tool in self.tools.values():
+            self.interpreter(inspect.getsource(tool.func))
```

Comment:
shall we move it to __init__, and only run this line for call-time tools? (the new change in ReAct)

Reply:
Since python interpreter is shutdown every `forward`, this needs to happen in `forward` too.